### PR TITLE
Use single ResizeObserver for better performance

### PR DESCRIPTION
**Pull request by xjmdoo**
According to the guidelines of using the `ResizeObserver` API the recommended approach is to use a single observer to observe multiple elements because it provides significant performance gains compared to creating a new observer for each element.

### CODE REVIEW

File: tests/integration/modifiers/did-resize-test.js [hunk 1]
```diff
@@ -87,7 +81,6 @@ module('Integration | Modifier | did-resize', function (hooks) {
 
     await render(hbs`<div {{did-resize this.resizeStub}}></div>`);
 
-    assert.notOk(resizeCallback, 'no callback received');
```
Comment (xjmdoo) [inline tests/integration/modifiers/did-resize-test.js:90]: I had to remove this `assert` since the observer is now `static` and `resizeCallback` might be set by the other test cases. I don't think this would be an issue as the next `assert` is the important one imo.

**Comment by elwayman02** [comment]
I don't think this change accomplishes the goal you set out to solve. Since modifiers are not singletons, and there is only ever one handler per modifier, then the code as written will always have 1 observer per 1 handler per 1 modifier. This PR does not change that, as the cached observer still only exists within the context of each separate modifier, which do not share state between them.

It sounds like what you actually want is a centralized singleton service to manage all ResizeObserver events from a single observer attached to many elements, akin to what we do for IntersectionObservers:

https://github.com/snewcomer/intersection-observer-admin
https://github.com/elwayman02/ember-scroll-modifiers/blob/master/addon/services/observer-manager.js

Does that seem correct or have I misunderstood what you're trying to accomplish?

**Comment by xjmdoo** [comment]
> I don't think this change accomplishes the goal you set out to solve. Since modifiers are not singletons, and there is only ever one handler per modifier, then the code as written will always have 1 observer per 1 handler per 1 modifier. This PR does not change that, as the cached observer still only exists within the context of each separate modifier, which do not share state between them.
> 
> It sounds like what you actually want is a centralized singleton service to manage all ResizeObserver events from a single observer attached to many elements, akin to what we do for IntersectionObservers:
> 
> https://github.com/snewcomer/intersection-observer-admin
> https://github.com/elwayman02/ember-scroll-modifiers/blob/master/addon/services/observer-manager.js
> 
> Does that seem correct or have I misunderstood what you're trying to accomplish?

Not sure if I understand your point, `static` properties on a class are shared between instances, because they are part of the class not the instance. I can easily miss something ember specific here but after testing, it was working and adding multiple modifiers to the page reused the observer and the one callback function managed all the handlers.

**Comment by elwayman02** [comment]
@xjmdoo Sorry for the lack of clarity in my initial response - it took me a bit to fully process your change and I edited my response above to be a bit more accurate. Generally, I am not sure if static class properties are the right pattern for this. I'm not saying they're not, but it's not something I've given a lot of consideration to. It's a novel approach and I want to noodle on it a bit and get some feedback from others before we move forward. It may well be that what you've done here is the best approach and it just wasn't something that occurred to me, in which case that's great!

I'm also adding this to the PR as a reference for others and justification for why this change is important:

https://github.com/WICG/resize-observer/issues/59#issuecomment-408098151
https://groups.google.com/a/chromium.org/g/blink-dev/c/z6ienONUb5A/m/F5-VcUZtBAAJ?pli=1

TLDR tests show that sharing a single ResizeObserver can be up to 10x faster than having a 1:1 relationship between elements and ResizeObservers. That's a pretty big win!

**Comment by xjmdoo** [comment]
@elwayman02 Thank you! Looking forward to your feedback. :)

File: tests/integration/modifiers/did-resize-test.js (hunk 1 above)
Comment (elwayman02) [reply tests/integration/modifiers/did-resize-test.js:90]: Wouldn't leaving `resizeCallback = null` in the `beforeEach` eliminate this issue by ensuring it gets reset before every test?

Comment (xjmdoo) [reply tests/integration/modifiers/did-resize-test.js:90]: If we do that, I think the issue is that it won't be set back to the callback because only the test case that runs first will create a new instance of the mock observer, after that it is shared by caching it using the static property.

Comment (elwayman02) [reply tests/integration/modifiers/did-resize-test.js:90]: You're right! I was mixing up the Observer and the Modifier itself here.

**Comment by elwayman02** [comment]
Ok! I'm back after discussing it with a few folks, and while services are a really comfortable approach, we think it's mostly 50/50 with either approach being valid. In this case, there are a couple things we kept in mind:

1) services are a more well-known Ember pattern, but static class properties are a native pattern and thus easier to teach even outside of Ember.
2) services are maybe slightly easier to test, but regardless either approach will require some knowledge of the internal implementation in order to properly test usage as a consumer
3) teardown with your approach is dead simple, as each modifier can simply remove itself when it's being torn down

So ultimately, we're fine moving forward with this approach! Thanks for bearing with me as I worked through that. I'll take a look through the PR again to just review for any issues with the implementation now that we're aligned on using the static class properties!

File: addon/modifiers/did-resize.js [hunk 2]
```diff
@@ -6,23 +6,41 @@ export default class DidResizeModifier extends Modifier {
...
+  addHandler() {
+    DidResizeModifier.handlers.push({
+      element: this.element,
+      handler: this.handler,
+    });
+  }
+
+  removeHandler() {
+    DidResizeModifier.handlers.splice(
+      DidResizeModifier.handlers.findIndex((h) => h.element === this.element),
```
Comment (elwayman02) [inline addon/modifiers/did-resize.js:22]: Since we're using `this.element` as the lookup key, would it makes sense to use a WeakMap here instead where the DOM element itself is the key and the handler is the value?

File: addon/modifiers/did-resize.js [hunk 3]
```diff
@@ -6,23 +6,41 @@ export default class DidResizeModifier extends Modifier {
...
+      this.addHandler();
+      DidResizeModifier.observer.observe(this.element, this.options);
     }
   }
 
   unobserve() {
-    if (this.observer) {
-      this.observer.unobserve();
+    if (DidResizeModifier.observer) {
+      DidResizeModifier.observer.unobserve(this.element);
```
Comment (elwayman02) [inline addon/modifiers/did-resize.js:36]: This method should probably also call `removeHandler`, since it's possible that the modifier's `handler` has been changed by the consumer.

File: addon/modifiers/did-resize.js [hunk 4]
```diff
@@ -6,23 +6,41 @@ export default class DidResizeModifier extends Modifier {
...
     }
   }
 
   disconnect() {
-    if (this.observer) {
-      this.observer.disconnect();
+    if (DidResizeModifier.observer) {
+      DidResizeModifier.observer.disconnect();
```
Comment (elwayman02) [inline addon/modifiers/did-resize.js:42]: I think we want to avoid calling `disconnect` altogether, because that unobserves ALL elements, now that they share a single observer instance. 

https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver/disconnect

In reality, it seems we could eliminate this method entirely and just call `unobserve` where we're calling `disconnect` on line 86.

File: addon/modifiers/did-resize.js [hunk 5]
```diff
@@ -38,19 +56,30 @@ export default class DidResizeModifier extends Modifier {
     this.handler = handler;
     this.options = options || this.options;
 
-    this.observe();
+    // Only observe if the modifier is installed
+    if (this._didInstall) {
```
Comment (elwayman02) [inline addon/modifiers/did-resize.js:60]: Is it possible to check that the `observer` property isn't null, instead of adding this boolean flag?

File: addon/modifiers/did-resize.js [hunk 6]
```diff
@@ -38,19 +56,30 @@ export default class DidResizeModifier extends Modifier {
...
     if (!('ResizeObserver' in window)) {
       return;
     }
 
-    this.observer = new ResizeObserver((entries, observer) => {
-      this.handler(entries[0], observer);
-    });
+    if (!DidResizeModifier.observer) {
+      DidResizeModifier.observer = new ResizeObserver((entries, observer) => {
+        for (let entry of entries) {
```
Comment (elwayman02) [inline addon/modifiers/did-resize.js:72]: This seems like a pretty expensive O(n^2) operation, having to iterate through all the entries and again through all of the handlers. Can this be simplified by using `this.element`?

```
for (let entry of entries) {
  if (this.element === entry.target) {
    this.handler(entry, observer);
  }
}
```

If not, maybe the proposed switch to WeakMap will simplify this logic somehow?

File: tests/integration/modifiers/did-resize-test.js [hunk 7]
```diff
@@ -1,40 +1,33 @@
...
-let unobserveStub;
-let disconnectStub;
-let MockResizeObserver;
-
 module('Integration | Modifier | did-resize', function (hooks) {
   setupRenderingTest(hooks);
 
-  let resizeObserver;
+  let resizeCallback = null;
+  const observeStub = sinon.stub();
```
Comment (elwayman02) [inline tests/integration/modifiers/did-resize-test.js:11]: I'd prefer to leave these as `let` for lines 11 through 15, since they are not conceptually being used as constants. I prescribe to the "prefer let" mode of thinking: https://madhatted.com/2016/1/25/let-it-be

File: tests/integration/modifiers/did-resize-test.js [hunk 8]
```diff
@@ -58,15 +51,16 @@ module('Integration | Modifier | did-resize', function (hooks) {
   });
 
   test('modifier triggers handler when ResizeObserver fires callback', async function (assert) {
-    await render(hbs`<div {{did-resize this.resizeStub}}></div>`);
-
-    let fakeEntry = { target: {} };
+    await render(
+      hbs`<div id="test-element" {{did-resize this.resizeStub}}></div>`
+    );
+    let entry = { target: find('#test-element') };
```
Comment (elwayman02) [inline tests/integration/modifiers/did-resize-test.js:57]: nice!

**Comment by elwayman02** [comment]
Overall, I think this looks great @xjmdoo! I left comments for a few minor tweaks/improvements we can consider, but this is in great shape and I think it won't take much at all to land this. Thank you for taking on such an important improvement for the addon; I'm glad you're finding it useful enough to contribute back to its development!

**Comment by elwayman02** [comment]
Closes #214

File: addon/modifiers/did-resize.js (hunk 5 above)
Comment (elwayman02) [reply addon/modifiers/did-resize.js:60]: This actually may be an opportunity to tackle #46 if you feel like taking that on. Moving the logic of `didInstall` to a constructor would eliminate the need to do this check at all. We would know that an observer exists because it *has* to be set up by the constructor of the first instance of this modifier, and we wouldn't call `this.observe` in the constructor, waiting instead to call it here within `didReceiveArguments`. What do you think?

**Comment by xjmdoo** [comment]
Thanks for the review @elwayman02 , all valid points. I've updated the PR with the changes and added some improvements to testing. Also, this should fix #46 now.

File: addon/modifiers/did-resize.js [hunk 9]
```diff
@@ -6,23 +6,46 @@ export default class DidResizeModifier extends Modifier {
...
+
+    if (!('ResizeObserver' in window)) {
+      return;
+    }
+
+    if (!DidResizeModifier.observer) {
+      DidResizeModifier.handlers = new WeakMap();
+      DidResizeModifier.observer = new ResizeObserver((entries, observer) => {
+        for (let entry of entries) {
+          const handler = DidResizeModifier.handlers.get(entry.target);
```
Comment (elwayman02) [inline addon/modifiers/did-resize.js:23]: I am confident this works as expected since this method will only get set up once, but since we're sharing this across all instances, is it possible to write a test that asserts each handler is only called once per observer event, when there are multiple modifier instances rendered? I just want to make sure that we have a test in place in case someone else comes along and changes this implementation in the future, since they might not immediately understand how this works.

**Comment by elwayman02** [comment]
Looks great! One more comment to see if we can write a reasonable test case just to make this a bit more solid. Otherwise, I think it's ready to go!

File: addon/modifiers/did-resize.js (hunk 9 above)
Comment (xjmdoo) [reply addon/modifiers/did-resize.js:23]: Good idea! I've added the new test case.

File: tests/integration/modifiers/did-resize-test.js [hunk 10]
```diff
@@ -102,4 +101,52 @@ module('Integration | Modifier | did-resize', function (hooks) {
     assert.ok(unobserveStub.calledOnce, 'unobserve called');
     assert.ok(observeStub.calledTwice, 'observe was called again');
   });
+
+  test('handlers are setup and called correctly when multiple modifiers are present', async function (assert) {
```
Comment (elwayman02) [inline tests/integration/modifiers/did-resize-test.js:105]: This test looks great! Nice work

**Comment by elwayman02** [comment]
Confirmed that the demo app still works in the preview, so I think we're good to go! Seeing as this is an enhancement that doesn't change the external API, I think we can release this as a minor version with no breaking changes. Thank you so much for this contribution @xjmdoo!
